src/dfx/src/util/mod.rs

```diff
@@ -215,6 +240,30 @@ pub fn blob_from_arguments(
     }
 }
 
+pub fn gather_principals_from_env(env: &dyn Environment) -> BTreeMap<String, String> {
+    use ic_agent::Identity;
+    let mut res: BTreeMap<String, String> = BTreeMap::new();
+    if let Ok(mgr) = env.new_identity_manager() {
+        let logger = env.get_logger();
+        if let Ok(names) = mgr.get_identity_names(logger) {
+            let mut map = names
+                .iter()
+                .filter_map(|name| {
+                    let id = mgr.load_identity(name, logger).ok()?;
```

Comment:
This should probably skip any encrypted identities, as well as HSM identities, and explicitly add `env.get_selected_identity_principal()`.

This was result when I ran `dfx new candid_assist && cd candid_assist && dfx deploy`:

```
The UI canister on the "local" network is "b77ix-eeaaa-aaaaa-qaada-cai"
Installing code for canister candid_assist_backend, with canister ID br5f7-7uaaa-aaaaa-qaaca-cai
Please enter the passphrase for your identity: 
```

My default identity doesn't have a passphrase.

After I entered a passphrase, on a later identity, it popped up the keychain dialog.

This will also potentially be an issue for HSM identities.  

Maybe we should store the identity's principal in the identity.json file for each identity.

Reply:
Good point. Is there a method to know which identities are encrypted? This is a best-effort function, we can start by including only the non-encrypted identities, or just the canister id from dfx.json, and gradually add more principals if people find it useful.

Reply:
Not exactly.  `IdentityManager.get_identity_config_or_default` is probably the closest.  I don't think it will do the right thing for the anonymous identity, which will be included in the results of get_identity_names().

Reply:
You'd want to look for identities with either:
- no identity.json
- load identity.json as IdentityConfiguration and all members are None